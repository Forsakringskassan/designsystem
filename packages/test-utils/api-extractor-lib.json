{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/lib.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/lib.d.ts"
    }
}
