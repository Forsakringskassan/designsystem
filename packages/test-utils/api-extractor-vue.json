{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/temp/types/vue/index.d.ts",
    "apiReport": {
        "enabled": false
    },
    "dtsRollup": {
        "publicTrimmedFilePath": "<projectFolder>/dist/types/vue.d.ts"
    }
}
