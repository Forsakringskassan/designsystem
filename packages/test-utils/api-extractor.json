{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../api-extractor.json",
    "tsdocMetadata": {
        "tsdocMetadataFilePath": "<projectFolder>/dist/types/tsdoc-metadata.json"
    },
    "dtsRollup": {
        "enabled": false
    }
}
