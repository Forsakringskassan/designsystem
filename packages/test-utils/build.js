const fs = require("node:fs/promises");
const path = require("node:path");
const esbuild = require("esbuild");

const type = {
    cjs: "commonjs",
    esm: "module",
};

async function build() {
    for (const format of ["cjs", "esm"]) {
        const outdir = `dist/${format}`;
        const common = {
            logLevel: "info",
            sourcemap: true,
            external: ["vue", "@fkui/*"],
            bundle: true,
            platform: "node",
            format,
            target: "node20",
            outdir,
        };
        await esbuild.build({
            ...common,
            entryPoints: ["src/lib.ts", "src/jest.ts"],
        });
        await esbuild.build({
            ...common,
            entryPoints: [{ in: "src/vue/index.ts", out: "vue" }],
            define: {
                "process.env.NODE_ENV": JSON.stringify("development"),
            },
        });
        const pkg = JSON.stringify({ type: type[format] }, null, 2);
        await fs.writeFile(path.join(outdir, "package.json"), pkg, "utf-8");
    }
}

build().catch((err) => {
    console.error(err);
});
