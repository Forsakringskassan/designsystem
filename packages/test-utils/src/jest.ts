/* imported for side-effects only */
import "./matchers";

export * from "./lib";
