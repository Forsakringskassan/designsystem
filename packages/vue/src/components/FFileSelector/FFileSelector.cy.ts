import { type DefineComponent, defineComponent } from "vue";
import { FFileSelectorPageObject } from "../../cypress";
import FFileSelector from "./FFileSelector.vue";

function createComponent(template: string): DefineComponent {
    return defineComponent({
        template,
        name: "FFileSelectorEnabled",
        components: { FFileSelector },
    });
}

describe("FFileSelector", () => {
    const fileSelectorEnabled = new FFileSelectorPageObject(
        '[data-test="enabledFileSelector"]',
    );
    const fileSelectorDisabled = new FFileSelectorPageObject(
        '[data-test="disabledFileSelector"]',
    );

    it("enabled fileselector should be visible, clickable and have a paper-clip icon", () => {
        const template = /* HTML */ `
            <f-file-selector
                v-test="'enabledFileSelector'"
                accept="application/pdf, image/jpeg, image/tiff, image/png"
            >
                Lägg till fil
            </f-file-selector>
        `;
        cy.mount(createComponent(template));
        fileSelectorEnabled.el().should("be.visible").click();
        fileSelectorEnabled.icon().should("be.visible");
        fileSelectorEnabled.addFile().should("not.have.attr", "aria-disabled");
    });

    it("disabled fileselector should be disabled", () => {
        const template = /* HTML */ `
            <f-file-selector
                id="dis"
                v-test="'disabledFileSelector'"
                accept="image/jpeg, image/tiff"
                disabled
            >
                Lägg till fil
            </f-file-selector>
        `;
        cy.mount(createComponent(template));
        fileSelectorDisabled
            .addFile()
            .should("have.attr", "aria-disabled", "true");
    });
});
