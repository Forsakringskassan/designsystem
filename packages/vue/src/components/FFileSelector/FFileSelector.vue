<script lang="ts">
import { defineComponent } from "vue";
import { ElementIdService } from "@fkui/logic";
import FIcon from "../FIcon/FIcon.vue";

export default defineComponent({
    name: "FFileSelector",
    components: {
        FIcon,
    },
    inheritAttrs: false,
    model: {
        prop: "files",
        event: "change",
    },
    props: {
        /**
         * The id for the input id attribute.
         * The id for the label for attribute.
         * If the prop is not set a random value will be generated.
         */
        id: {
            type: String,
            required: false,
            default: () => ElementIdService.generateElementId(),
        },
        /**
         * Disables the file selector.
         */
        disabled: {
            type: Boolean,
            required: false,
            default: false,
        },
    },
    emits: ["change"],
    computed: {
        attrs(): Record<string, unknown> {
            return {
                ...this.$attrs,
                id: this.id,
                onChange: (event: Event) => {
                    if (event.target instanceof HTMLInputElement) {
                        /**
                         * V-model event.
                         *
                         * @event input
                         * @type {FileList}
                         */
                        this.$emit("change", event.target.files);
                    }
                },
            };
        },

        labelClass(): string {
            return this.disabled ? "disabled" : "enabled";
        },
        labelId(): string {
            return `${this.id}_label`;
        },
        ariaDisabled(): string | undefined {
            return this.disabled ? "true" : undefined;
        },
    },
    methods: {
        onClick(event: Event) {
            if (this.disabled) {
                event.preventDefault();
                return;
            }

            const input = this.$refs["file-selector"] as HTMLInputElement;
            input.value = "";
        },
    },
});
</script>

<template>
    <div class="file-selector">
        <input
            :id="id"
            ref="file-selector"
            type="file"
            :aria-labelledby="labelId"
            :aria-disabled="ariaDisabled ? 'true' : undefined"
            v-bind="attrs"
            @click="onClick"
        />
        <label
            :id="labelId"
            role="button"
            :class="labelClass"
            :for="id"
            class="button button--tertiary button--medium"
            aria-hidden="true"
        >
            <f-icon class="button__icon" name="paper-clip"></f-icon>
            <slot></slot>
        </label>
    </div>
</template>
