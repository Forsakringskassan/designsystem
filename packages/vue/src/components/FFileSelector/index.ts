export { default as FFileSelector } from "./FFileSelector.vue";
