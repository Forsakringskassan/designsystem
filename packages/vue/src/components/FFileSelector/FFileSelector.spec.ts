import "html-validate/jest";
import { mount, VueWrapper } from "@vue/test-utils";
import FFileSelector from "./FFileSelector.vue";

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
} = {}): VueWrapper {
    return mount(FFileSelector, {
        attrs: { ...attrs },
        props: { ...props },
        slots: { default: "Select file", ...slots },
        global: {
            stubs: ["FIcon"],
        },
    });
}

describe("FFileselector", () => {
    it("should pass attributes", () => {
        const filetypesAccepted =
            "application/pdf, image/jpeg, image/tiff, image/png";
        const wrapper = createWrapper({
            attrs: {
                accept: filetypesAccepted,
            },
        });
        const input = wrapper.get("input");
        expect(input.attributes("accept")).toBe(filetypesAccepted);
    });

    it("should set text in slot", () => {
        const wrapper = createWrapper();
        const label = wrapper.get("label");
        expect(label.text()).toBe("Select file");
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should pass listeners", async () => {
        const foobar = jest.fn();
        const wrapper = createWrapper({
            attrs: { onFoobar: foobar },
        });
        const element = wrapper.get("input");
        await element.trigger("foobar");
        expect(foobar).toHaveBeenCalled();
    });

    it("should emit on change", async () => {
        const wrapper = createWrapper();
        const input = wrapper.get("input");
        await input.trigger("change");
        // emits Event and FileList on change
        expect(wrapper.emitted("change")).toHaveLength(1);
    });

    it("should disable component when disabled", () => {
        const wrapper = createWrapper({
            props: { disabled: true },
        });
        const input = wrapper.get("input");
        const label = wrapper.get("label");
        expect(input.attributes("aria-disabled")).toBe("true");
        expect(label.classes()).toContain("disabled");
    });

    it("should not disable component when not disabled", () => {
        const wrapper = createWrapper();
        const input = wrapper.get("input");
        const label = wrapper.get("label");
        expect(input.attributes("aria-disabled")).toBeUndefined();
        expect(label.classes()).not.toContain("disabled");
    });

    /*
     * Fixes bug related to WCAG 2.5.3 (label in name).
     * The file selector's visible text did not match the text that was read aloud by screen reader.
     * Instead, the screen reader read the content of the browser's default input[file]
     * (e.g. "VÃ¤lj fil: Ingen fil vald" in Edge).
     */
    it("should have input with aria-labelledby attribute", () => {
        const wrapper = createWrapper({
            props: { id: "foo" },
        });
        const input = wrapper.get("input");
        const label = wrapper.get("label");
        expect(label.attributes("id")).toBe("foo_label");
        expect(input.attributes("aria-labelledby")).toBe("foo_label");
    });

    /*
     * Fixes SFKUI-683. By using the keyboard's arrow keys to navigate past the file selector, the file selector
     * was read out twice by screen reader.
     */
    it("should have label with aria-hidden attribute", () => {
        const wrapper = createWrapper();
        const label = wrapper.get("label");
        expect(label.attributes("aria-hidden")).toBe("true");
    });
});

describe("html-validate", () => {
    describe("attributes", () => {
        it("disabled", () => {
            expect.assertions(2);
            const valid = /* HTML */ `
                <f-file-selector>text</f-file-selector>
                <f-file-selector disabled>text</f-file-selector>
                <f-file-selector :disabled="false">text</f-file-selector>
                <f-file-selector :disabled="true">text</f-file-selector>
            `;
            const invalid = /* HTML */ `
                <f-file-selector disabled="foobar">text</f-file-selector>
            `;
            expect(valid).toMatchInlineCodeframe(`""`);
            expect(invalid).toMatchInlineCodeframe(`
                "error: Attribute "disabled" should omit value (attribute-boolean-style) at inline:2:34:
                  1 |
                > 2 |                 <f-file-selector disabled="foobar">text</f-file-selector>
                    |                                  ^^^^^^^^
                  3 |
                Selector: f-file-selector
                error: Attribute "disabled" has invalid value "foobar" (attribute-allowed-values) at inline:2:44:
                  1 |
                > 2 |                 <f-file-selector disabled="foobar">text</f-file-selector>
                    |                                            ^^^^^^
                  3 |
                Selector: f-file-selector"
            `);
        });
    });
});
