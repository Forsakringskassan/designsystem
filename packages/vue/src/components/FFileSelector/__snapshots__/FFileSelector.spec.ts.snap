// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FFileselector should set text in slot 1`] = `
<div
  class="file-selector"
>
  <input
    aria-labelledby="fkui-vue-element-0001_label"
    id="fkui-vue-element-0001"
    type="file"
  />
  <label
    aria-hidden="true"
    class="enabled button button--tertiary button--medium"
    for="fkui-vue-element-0001"
    id="fkui-vue-element-0001_label"
    role="button"
  >
    <f-icon-stub
      class="button__icon"
      library="f"
      name="paper-clip"
    />
    
    Select file
    
  </label>
</div>
`;
